accumulator_updater/programs/accumulator_updater/src/instructions/put_all.rs

```diff
@@ -26,85 +26,58 @@ pub fn put_all<'info>(
-        let bump = if is_uninitialized_account(ai) {
-            let seeds = &[
+        let accumulator_input = &mut (if is_uninitialized_account(accumulator_input_ai) {
+            let (pda, bump) = Pubkey::find_program_address(
+                &[
+                    cpi_caller.as_ref(),
+                    b"accumulator".as_ref(),
+                    base_account_key.as_ref(),
+                ],
+                &crate::ID,
+            );
+            require_keys_eq!(accumulator_input_ai.key(), pda);
+            let signer_seeds = &[
                 cpi_caller.as_ref(),
                 b"accumulator".as_ref(),
                 base_account_key.as_ref(),
-                &account_schema.to_le_bytes(),
+                &[bump],
             ];
-            let (pda, bump) = Pubkey::find_program_address(seeds, &crate::ID);
-            require_keys_eq!(ai.key(), pda);
-
-
-            //TODO: Update this with serialization logic
-            // 8 for anchor discriminator
-            let accumulator_size = 8 + AccumulatorInput::size(&account_data);
             PutAll::create_account(
-                ai,
-                accumulator_size,
+                accumulator_input_ai,
+                8 + AccumulatorInput::INIT_SPACE,
                 &ctx.accounts.payer,
-                &[
-                    cpi_caller.as_ref(),
-                    b"accumulator".as_ref(),
-                    base_account_key.as_ref(),
-                    &account_schema.to_le_bytes(),
-                    &[bump],
-                ],
-                &rent,
+                signer_seeds,
                 &ctx.accounts.system_program,
             )?;
-            initialized.push(ai.key());
-
-            bump
-        } else {
-            let accumulator_input = <AccumulatorInput as AccountDeserialize>::try_deserialize(
-                &mut &**ai.try_borrow_mut_data()?,
+            loader = AccountLoader::<AccumulatorInput>::try_from_unchecked(
+                &crate::ID,
+                accumulator_input_ai,
             )?;
-            {
-                // TODO: allow re-sizing?
-                require_gte!(
-                    accumulator_input.data.len(),
-                    account_data.len(),
-                    AccumulatorUpdaterError::CurrentDataLengthExceeded
-                );
-
-                accumulator_input.validate(
-                    ai.key(),
-                    cpi_caller,
-                    base_account_key,
-                    account_schema,
-                )?;
-            }
-
+            let mut accumulator_input = loader.load_init()?;
+            accumulator_input.header = AccumulatorHeader::new(bump);
+            accumulator_input
+        } else {
+            loader = AccountLoader::<AccumulatorInput>::try_from(accumulator_input_ai)?;
+            let mut accumulator_input = loader.load_mut()?;
+            accumulator_input.validate(accumulator_input_ai.key(), cpi_caller, base_account_key)?;
```

Comment:
I would suggest putting this `validate` call outside the if/else right above `put_all`, as I think that reduces the chance for bugs. (It's redundant for the first codepath, but it makes it easier to reason that the code performs the necessary security checks.)